File: windows/OpenMeKit.Tests/KnockServiceTests.cs
using OpenMeKit;
using Xunit;

namespace OpenMeKit.Tests;

/// <summary>Unit tests for <see cref="KnockService"/> packet construction.</summary>
public sealed class KnockServiceTests
{
    // A pair of real test keys (ed25519 client + x25519 server) suitable for unit tests.
    // These are NOT used in production; they are committed only for test purposes.
    private const string TestServerPubKey  = "wfLHOBMHSXQT5YN0yCFW2y0TKhkXPrfMjklBkdGg5kM="; // 32 B X25519
    private const string TestClientPrivKey = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";   // 32 B Ed25519 seed (all-zeros test vector)

    [Fact]
    public void BuildPacket_ReturnsExact165Bytes()
    {
        var packet = KnockService.BuildPacket(TestServerPubKey, TestClientPrivKey);
        Assert.Equal(KnockService.PacketSize, packet.Length);
    }

    [Fact]
    public void BuildPacket_FirstByteIsProtocolVersion()
    {
        var packet = KnockService.BuildPacket(TestServerPubKey, TestClientPrivKey);
        Assert.Equal(KnockService.ProtocolVersion, packet[0]);
    }

    [Fact]
    public void BuildPacket_TwoCallsDifferInEphemeralKey()
    {
        // Ephemeral key (bytes 1-32) must be freshly generated each time.
        var p1 = KnockService.BuildPacket(TestServerPubKey, TestClientPrivKey);
        var p2 = KnockService.BuildPacket(TestServerPubKey, TestClientPrivKey);
        Assert.NotEqual(p1[1..33], p2[1..33]);
    }

    [Fact]
    public void BuildPacket_InvalidServerKeyThrows()
    {
        var ex = Assert.Throws<KnockException>(() =>
            KnockService.BuildPacket("not-base64!!", TestClientPrivKey));
        Assert.Equal(KnockErrorKind.InvalidServerKey, ex.Kind);
    }

    [Fact]
    public void BuildPacket_WrongLengthServerKeyThrows()
    {
        // 16 bytes → not 32
        var ex = Assert.Throws<KnockException>(() =>
            KnockService.BuildPacket(Convert.ToBase64String(new byte[16]), TestClientPrivKey));
        Assert.Equal(KnockErrorKind.InvalidServerKey, ex.Kind);
    }

    [Fact]
    public void BuildPacket_InvalidClientKeyThrows()
    {
        var ex = Assert.Throws<KnockException>(() =>
            KnockService.BuildPacket(TestServerPubKey, "!!!invalid!!!"));
        Assert.Equal(KnockErrorKind.InvalidClientKey, ex.Kind);
    }

    [Fact]
    public void BuildPacket_64ByteClientKeyAccepted()
    {
        // 64-byte key = seed(32) + pub(32); only the seed should be used.
        var seed = new byte[32];
        var pub  = new byte[32];
        var combined = Convert.ToBase64String([.. seed, .. pub]);

        var packet = KnockService.BuildPacket(TestServerPubKey, combined);
        Assert.Equal(KnockService.PacketSize, packet.Length);
    }
}

File: apple/OpenMeKit/Tests/OpenMeKitTests/KnockServiceTests.swift
import XCTest
import CryptoKit
@testable import OpenMeKit

/// Unit tests for ``KnockService`` packet construction and key validation.
///
/// These tests exercise the pure packet-building path (no UDP transmission).
/// They validate the wire-format invariants described at
/// https://openme.merlos.org/docs/protocol/packet-format.html
final class KnockServiceTests: XCTestCase {

    // MARK: - Fixtures

    /// A freshly generated server key pair used across tests.
    private let serverPriv = Curve25519.KeyAgreement.PrivateKey()
    private var serverPub:  Curve25519.KeyAgreement.PublicKey { serverPriv.publicKey }

    /// A freshly generated client signing key used across tests.
    private let signingKey = Curve25519.Signing.PrivateKey()

    // MARK: - Packet size

    func testPacketIsExactly165Bytes() throws {
        let packet = try KnockService.buildPacket(serverPubKey: serverPub, signingKey: signingKey)
        XCTAssertEqual(packet.count, KnockService.packetSize,
                       "buildPacket must return exactly \(KnockService.packetSize) bytes")
    }

    // MARK: - Protocol version

    func testFirstByteIsProtocolVersion() throws {
        let packet = try KnockService.buildPacket(serverPubKey: serverPub, signingKey: signingKey)
        XCTAssertEqual(packet[0], KnockService.protocolVersion,
                       "First byte must be protocol version \(KnockService.protocolVersion)")
    }

    // MARK: - Ephemeral randomness

    func testTwoKnocksProduceDifferentPackets() throws {
        let p1 = try KnockService.buildPacket(serverPubKey: serverPub, signingKey: signingKey)
        let p2 = try KnockService.buildPacket(serverPubKey: serverPub, signingKey: signingKey)
        XCTAssertNotEqual(p1, p2,
                          "Each knock must embed a fresh ephemeral key and random nonce")
    }

    func testEphemeralPublicKeyDiffersPerKnock() throws {
        let p1 = try KnockService.buildPacket(serverPubKey: serverPub, signingKey: signingKey)
        let p2 = try KnockService.buildPacket(serverPubKey: serverPub, signingKey: signingKey)
        // Bytes 1-32 (inclusive) are the ephemeral Curve25519 public key
        XCTAssertNotEqual(p1[1..<33], p2[1..<33],
                          "Ephemeral public key (bytes 1-32) must be fresh per knock")
    }

    // MARK: - Field layout

    func testSignedPortionIs101Bytes() {
        XCTAssertEqual(KnockService.signedPortionSize, 101)
    }

    func testPacketSizeConstant() {
        XCTAssertEqual(KnockService.packetSize, 165)
    }

    func testProtocolVersionConstant() {
        XCTAssertEqual(KnockService.protocolVersion, 1)
    }

    // MARK: - Key validation (via knock API)

    func testInvalidBase64ServerKeyFails() {
        let expectation = expectation(description: "invalid server key fails")
        KnockService.knock(
            serverHost: "127.0.0.1",
            serverPort: 54154,
            serverPubKeyBase64: "not-base64!!!",
            clientPrivKeyBase64: Data(signingKey.rawRepresentation).base64EncodedString()
        ) { result in
            if case .failure(let err) = result, case .invalidServerKey = err {
                // expected
            } else {
                XCTFail("Expected invalidServerKey error, got \(result)")
            }
            expectation.fulfill()
        }
        waitForExpectations(timeout: 2)
    }

    func testWrongLengthServerKeyFails() {
        let expectation = expectation(description: "wrong-length server key fails")
        let shortKey = Data(repeating: 0, count: 16).base64EncodedString()   // 16 B, not 32
        KnockService.knock(
            serverHost: "127.0.0.1",
            serverPort: 54154,
            serverPubKeyBase64: shortKey,
            clientPrivKeyBase64: Data(signingKey.rawRepresentation).base64EncodedString()
        ) { result in
            if case .failure(let err) = result, case .invalidServerKey = err {
                // expected
            } else {
                XCTFail("Expected invalidServerKey error, got \(result)")
            }
            expectation.fulfill()
        }
        waitForExpectations(timeout: 2)
    }

    func testInvalidBase64ClientKeyFails() {
        let expectation = expectation(description: "invalid client key fails")
        KnockService.knock(
            serverHost: "127.0.0.1",
            serverPort: 54154,
            serverPubKeyBase64: Data(serverPub.rawRepresentation).base64EncodedString(),
            clientPrivKeyBase64: "!!not-base64!!"
        ) { result in
            if case .failure(let err) = result, case .invalidClientKey = err {
                // expected
            } else {
                XCTFail("Expected invalidClientKey error, got \(result)")
            }
            expectation.fulfill()
        }
        waitForExpectations(timeout: 2)
    }

    // MARK: - Packet constant check via build (smoke test)

    func testBuildPacketProducesNonZeroBytes() throws {
        let packet = try KnockService.buildPacket(serverPubKey: serverPub, signingKey: signingKey)
        // The signed portion (bytes 0-100) must not be all zeros.
        let signedPortion = packet[0..<KnockService.signedPortionSize]
        let allZero = signedPortion.allSatisfy { $0 == 0 }
        XCTAssertFalse(allZero, "Signed portion must not be all zeros")
    }
}
